src/instruments/src/MCDU/Pages/FMGC/RadNav/Radnav.tsx

```diff
@@ -0,0 +1,381 @@
+        } else {
+            setValue(value);
+        }
+    };
+
+    const validateEntry = (value: string) => {
+        if (value === '') {
+            addMessage(NXSystemMessages.formatError);
+            return false;
+        }
+        return true;
+    };
+    return (
+        <LineHolder columnPosition={1}>
+            <LabelField lineSide={lineSides.left} value="ADF1/FREQ" color={lineColors.white} />
+            <StringInputField
+                lineSide={lineSides.left}
+                defaultValue={value}
+                nullValue="[  ]/[   .]"
+                color={value !== undefined ? lineColors.cyan : lineColors.cyan}
+                lsk={LINESELECT_KEYS.L5}
+                selectedCallback={(value) => setNewValue(value)}
+                selectedValidation={((value) => validateEntry(value))}
+                size={lineSizes.regular}
+            />
+        </LineHolder>
+    );
+};
+
+type AdftwoProps = {
+    addMessage: (msg: ScratchpadMessage) => void,
+}
+const Adftwo:React.FC<AdftwoProps> = ({ addMessage }) => {
+    const [value, setValue] = useState<string | undefined>(undefined);
+
+    const setNewValue = (value: string | undefined) => {
+        if (value === undefined) {
+            addMessage(NXSystemMessages.notAllowed);
+        } else {
+            setValue(value);
+        }
+    };
+
+    const validateEntry = (value: string) => {
+        if (value === '') {
+            addMessage(NXSystemMessages.formatError);
+            return false;
+        }
+        return true;
+    };
+    return (
+        <LineHolder columnPosition={2}>
+            <LabelField lineSide={lineSides.right} value="FREQ/ADF2" color={lineColors.white} />
+            <StringInputField
+                lineSide={lineSides.right}
+                defaultValue={value}
+                nullValue="[    .]/[ ]"
+                color={value !== undefined ? lineColors.cyan : lineColors.cyan}
+                lsk={LINESELECT_KEYS.R5}
+                selectedCallback={(value) => setNewValue(value)}
+                selectedValidation={((value) => validateEntry(value))}
+                size={lineSizes.regular}
+            />
+        </LineHolder>
+    );
+};
+
+const Adfonebfo:React.FC = () => (
+    <LineHolder columnPosition={1}>
+        <EmptyLine />
+        <Field
+            lineSide={lineSides.left}
+            value="<-ADF1 BFO"
+            size={lineSizes.regular}
+            color={lineColors.inop}
+        />
+    </LineHolder>
+);
+const Adftwobfo:React.FC = () => (
+    <LineHolder columnPosition={2}>
+        <EmptyLine />
+        <Field
+            lineSide={lineSides.right}
+            value="ADF2 BFO->"
+            size={lineSizes.regular}
+            color={lineColors.inop}
+        />
+    </LineHolder>
+);
+type RadnavePageProps = {
+    addMessage: (msg: ScratchpadMessage) => void,
+}
+
+const RadnavPage: React.FC<RadnavePageProps> = ({ addMessage }) => {
+    const dispatch = useMCDUDispatch();
```

Comment:
in reference to my other comment, you already have the dispatcher here so just do 
```js
import * as scratchpadActions from '../../../redux/actions/scratchpadActionCreators';
const addMessage = (msg: ScratchpadMessage) => {
        dispatch(scratchpadActions.addScratchpadMessage(msg));
};
```
See below for a usage example https://github.com/flybywiresim/a32nx/blob/1bdbc95860f8b482c0928e3e349256d74aa6ef84/src/instruments/src/MCDU/Pages/FMGC/Init/Init.tsx#L56-L58